### ISSUE TITLE
Suggestion: Always generate parent type for resolvers

### ISSUE THREAD

**Issue by jonaskello** [dotansimha/graphql-code-generator#1589]
**Is your feature request related to a problem? Please describe.**

Naming of types that is returned by resolvers is inconsistent.

**Describe the solution you'd like**

In order to set the context for this suggestion, let's define the concept of a `FieldResolver` and a `TypeResolver`. A `FieldResolver` resolves a field on a type, and a `TypeResolver` is a map of `FieldResolver`s.

To give an example, this is generated today by the `typescript-resolvers` plugin:

```ts
export type UserResolvers<Context = any, ParentType = User> = {
  id: Resolver<Scalars["String"], ParentType, Context>;
  name: Resolver<Scalars["String"], ParentType, Context>;
};
```

In this example `UserResolvers` is a `TypeResolver` and `Resolver<...>` is a `FieldResolver`. (A perhaps clearer but also longer naming would be `UserTypeResolver` and `FieldResolver<...>`)

A `TypeResolver` accepts a single `ParentType` which is usually a semi-resolved type that the `TypeResolver` will fully resolve. I think this means that the return type of all `FieldResolver`s is not random, it must always the be exact same type that the `TypeResolver` accepts as `ParentType`. So if we have some types in the schema that have fields that resolves to a `User` type, they all need to return the exact `ParentType` that the `User` type's `TypeResolver` will accept.

The main challange with typings for type-resolvers is how to generate the `ParentType = User` part. Today it will by default assign the full schema type which means all `FieldResolver`s needs to return a fully resolved type which probably is not what we want (as noted in #1219). There is a nice [mapping feature](https://graphql-code-generator.com/docs/plugins/typescript-resolvers#mappers-overwrite-parents-and-resolved-values) that allows us to change this. For example we can map `User: UserParent`. This means that the `User` type's `TypeResolver` will have `ParentType = UserParent`.

Now to my suggestion. When writing a new `FieldResolver` that returns schema-type `User` I need to think about what typescript-type to return. Something like this:

```ts
export const queryResolver: GQLResolvers.QueryResolvers<Context, RootValue> = {
  async users(
    parent: RootValue,
    args: GQLResolvers.QueryUsersArgs,
    ctx: Context
  ): Promise<ReadonlyArray<SOMETHING>> {
      // Resolve and return "SOMETHING"
  },
};
```
So the challange here is to know what to put in "SOMETHING". Yes, if I would leave out all the type annotations it would be infrared. However I like to annotate all function inputs and return value and have a linting rule that enforces this. So now that I'm writing this resolver the question becomes, should I return the full `User` type or is there perpaps a mapping for a the return type? In order to know this I need to do a lookup, either check the generated types or check the mapping config for the code generator. The reason for this lookup is that the return type for `FieldResolver`s is not consistent. Sometimes it is the full schema type and sometimes it is something else. This is the way it has to be, however I think it would be nice to have the *naming* of these types consistent. So my suggestion is that there is always a type-alias generated for the `ParentType` of a `TypeResolver`. For example:

```ts
type UserParent = User; // This is a generated type alias

export type UserResolvers<Context = any, ParentType = UserParent> = {
  id: Resolver<Scalars["String"], ParentType, Context>;
  name: Resolver<Scalars["String"], ParentType, Context>;
};
```

By default the type alias would be the same as the schema type so it does not seem to add much value. However, now I always know the name of the type to return from `FieldResolver`s without doing a lookup. If the `FieldResolver` should return the schema type `User` then I by convention know that it should return the typescript type `UserParent`. Another benefit of this is that if I later want to change the type that the `TypeResolver` for `User` accepts (by adding a mapping `User`->`UserSpecial`), I don't need to go back and refactor all my type annotations from `User` to `UserSpecial`. The code generator could just map the type-alias to `UserParent`->`UserSpecial` and all annotations in existing `FieldResolver`s return types could remain the same (they are always consistently refering to `UserParent`).

Another idea would be to use the same approach that is used for Scalars. The generated code would be something like this:

```ts
export type Parents = {
  User: User;
  // Other parents
};

export type UserResolvers<Context = any, ParentType = Parents["User"]> = {
  id: Resolver<Scalars["String"], ParentType, Context>;
  name: Resolver<Scalars["String"], ParentType, Context>;
};
```

Now overriding the parent type would use the same approach as overriding a scalar (just affecting the `Parents` mapping and nothing else).

I hope this makes sense and you can follow this long explaination of a rather simple suggestion :-).  I'm not sure if this approach is good but at least in my case I think it would make my code cleaner and easier to understand (by consistently having the same return type naming for field resolvers).

**Comment by dotansimha** [comment]
Hi @jonaskello , 
Thank you for your suggestion. 
We are working now on improving the resolvers packages, we want to improve the usage of mappers, by overriding the inner types and links when mappers are used. It also requires to create a `ResolverTypes` type that contains mapping between GraphQL type to a TypeScript type (which is the same as your suggested, but it will also work for return types).

**Comment by dotansimha** [comment]
Implemented in 1.0.7 :)
